### ISSUE TITLE
Access token not renewed if expired

### ISSUE THREAD

**Issue by klues** [authts/oidc-client-ts#1601]
Silent renew of the access token [is only done if it expires soon, not if it's already expired](https://github.com/authts/oidc-client-ts/blob/main/src/SilentRenewService.ts#L23). It's possible to manually trigger renewal for expired access tokens by
```
userManager.events.addAccessTokenExpired(function() {
     userManager.signinSilent();
 })
```
but I wonder, why it's not done by default? Any specific reason for this, or could it be added by default?

**Comment by pamapa** [comment]
`userManager.getUser()` should trigger an renew...

**Comment by klues** [comment]
OK, but it doesn't do it, if the access token is already expired, only if it's expiring soon. That's my question if this should be changed?!

**Comment by pamapa** [comment]
> OK, but it doesn't do it, if the access token is already expired, only if it's expiring soon. That's my question if this should be changed?!

When a look at the code, it does (see "if it's negative, it will still fire" below):
https://github.com/authts/oidc-client-ts/blob/1aaeb9441a3235414e78d9f0e889712e4ec88263/src/AccessTokenEvents.ts#L26-L57

You may need to enable logging to console at debugging level to see what is going on with your application:
```
import { Log } from "oidc-client-ts";
Log.setLogger(console);
Log.setLevel(Log.DEBUG)
```

**Comment by klues** [comment]
Yes, the event `AccessTokenExpired` is fired, but `userManager.getUser()` doesn't renew the tokens, if the access token is already expired. Therefore it's mandatory to add the code snippet I've posted to manually run `userManager.signinSilent` if the access token is expired.

**Comment by pamapa** [comment]
> Yes, the event `AccessTokenExpired` is fired, but `userManager.getUser()` doesn't renew the tokens, if the access token is already expired. Therefore it's mandatory to add the code snippet I've posted to manually run `userManager.signinSilent` if the access token is expired.

Have you enabled `automaticSilentRenew`? This code is expected to run when "access token expired" happened:
https://github.com/authts/oidc-client-ts/blob/1aaeb9441a3235414e78d9f0e889712e4ec88263/src/SilentRenewService.ts#L47-L63

**Comment by pamapa** [comment]
> I assume that the problem is that it's only listening to `AccessTokenExpiring` and not also `AccessTokenExpired` here

I agree. This code of only doing `addAccessTokenExpiring` is very old. I guess we can simply extend that and do that as well `addAccessTokenExpired` after the first register and also we need to remove it again. Can you try this out on your side?

**Comment by pamapa** [comment]
This is probably a better way to handle this within `SilentRenewService.ts`:
```ts
...
            // this will trigger loading of the user so the expiring events can be initialized
            try {
                const user = await this._userManager.getUser();
                if (user?.expired) {
                    // we only listen on token expiring, thus request a new token if the current one is expired
                    await this._tokenExpiring();
                }
            }
            catch (err) {
                // catch to suppress errors since we're in a ctor
                logger.error("getUser error", err);
            }
...
```
Can you test this?

**Comment by MFredX** [comment]
@franck102 in your case, is it possible to redirect the user to authorisation endpoint to reauthenticate?
I feel it is difficult to suggest what best course of action without understanding what exactly causes the `signinSilent` to fail.

@pamapa 
What is the usual process when testing changes like that? I would like to help with the testing as I am affected by the problem as well
